services-gateway-netty/src/main/java/io/scalecube/services/gateway/GatewaySessionHandler.java

```diff
@@ -21,42 +22,77 @@
    * @param req request message (not null)
    * @return message
    */
-  default M mapMessage(GatewaySession session, M req) {
+  default M mapMessage(GatewaySession session, M req, Context context) {
     return req;
   }
 
+  /**
+   * Request mapper function.
+   *
+   * @param session session
+   * @param byteBuf request buffer
+   * @param context subscriber context
+   * @return context
+   */
+  default Context onRequest(GatewaySession session, ByteBuf byteBuf, Context context) {
+    return context;
+  }
+
+  /**
+   * On response handler.
+   *
+   * @param session session
+   * @param byteBuf response buffer
+   * @param message response message
+   * @param context subscriber context
+   */
+  default void onResponse(
+      GatewaySession session, ByteBuf byteBuf, GatewayMessage message, Context context) {
+    // no-op
+  }
+
   /**
    * Error handler function.
    *
    * @param session webscoket session (not null)
    * @param throwable an exception that occurred (not null)
-   * @param req request message (optional)
-   * @param resp response message (optional)
+   * @param context subscriber context
    */
-  default void onError(GatewaySession session, Throwable throwable, M req, M resp) {
+  default void onError(GatewaySession session, Throwable throwable, Context context) {
     LOGGER.error(
-        "Exception occurred on session={}, on request: {}, on response: {}, cause:",
+        "Exception occurred on session: {}, on context: {}, cause:",
         session.sessionId(),
-        req,
-        resp,
+        context,
         throwable);
   }
 
+  /**
+   * On session error.
+   *
+   * @param session webscoket session (not null)
+   * @param throwable an exception that occurred (not null)
+   */
+  default void onSessionError(GatewaySession session, Throwable throwable) {
+    LOGGER.error("Exception occurred on session: {}, cause:", session.sessionId(), throwable);
+  }
+
   /**
    * On session open handler.
    *
    * @param session websocket session (not null)
+   * @return mono result
    */
-  default void onSessionOpen(GatewaySession session) {
-    LOGGER.info("Session opened: " + session);
+  default Mono<Void> onSessionOpen(GatewaySession session) {
+    return Mono.fromRunnable(() -> LOGGER.info("Session opened: {}", session));
   }
 
   /**
    * On session close handler.
    *
    * @param session websocket session (not null)
+   * @return mono result
    */
-  default void onSessionClose(GatewaySession session) {
-    LOGGER.info("Session closed: " + session);
+  default Mono<Void> onSessionClose(GatewaySession session) {
```

Comment:
Why do we need such changes? I mean what does using Mono<Void> give us? If you need to throw en exception you can just throw it, no need to wrap it to `Mono.error`

Reply:
You 're talking about onSessionClose specifically?

Reply:
@artem-v yes, it is

Reply:
I also was unsure about it and make it to be in sync with onSessionOpen. But to be fair u don't need to be async  on close.

Reply:
> @artem-v yes, it is, and about `onSessionOpen`

Im planning to something asynchronous during session-open phase, and don't want to block netty thread, hence this mono approach.

Reply:
But for `onSessionOpen` is redundant as well

Reply:
> Im planning to do something asynchronous during session-open phase, and don't want to block netty thread, hence this mono approach.

I thought it's already asynchronous.

Reply:
Yes
You're correct. 
I will refactor this change.

Reply:
@artem-v Thank you! My apologies for stopping merge this PR.